src/Database/Expression/WindowExpression.php

```diff
@@ -83,22 +94,102 @@ public function order($fields)
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::RANGE, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::RANGE, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
     }
 
     /**
      * @inheritDoc
      */
     public function rows(?int $start, ?int $end = 0)
     {
-        return $this;
+        if (func_num_args() === 1) {
+            return $this->frame(self::ROWS, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::ROWS, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
     }
 
     /**
      * @inheritDoc
      */
     public function groups(?int $start, ?int $end = 0)
     {
+        if (func_num_args() === 1) {
+            return $this->frame(self::GROUPS, $start, self::PRECEDING);
+        } else {
+            return $this->frame(self::GROUPS, $start, self::PRECEDING, $end, self::FOLLOWING);
+        }
+    }
+
+    /**
+     * @inheritDoc
+     */
+    public function frame(
+        int $type,
+        $startOffset,
+        int $startDirection,
+        $endOffset = null,
+        int $endDirection = self::FOLLOWING
+    ) {
+        $validTypes = [self::RANGE, self::ROWS, self::GROUPS];
+        if (!in_array($type, $validTypes)) {
+            throw new InvalidArgumentException('Frame type must be RANGE, ROWS or GROUP.');
+        }
+
+        $validDirections = [self::PRECEDING, self::FOLLOWING];
+        if (
+            !in_array($startDirection, $validDirections) ||
+            !in_array($endDirection, $validDirections)
+        ) {
+            throw new InvalidArgumentException('Frame directions must be PRECEDING or FOLLOWING.');
+        }
+
+        if (
+            (is_int($startOffset) && $startOffset < 0) ||
+            (is_int($endOffset) && $endOffset < 0)
+        ) {
+            throw new InvalidArgumentException('Frame offsets must be non-negative.');
+        }
+
+        if (
+            in_array($type, [self::ROWS, self::GROUPS]) &&
+            (
+                ($startOffset !== null && !is_int($startOffset)) ||
+                ($endOffset !== null && !is_int($endOffset))
+            )
+        ) {
+            throw new InvalidArgumentException('Frame offsets for ROWS and GROUPS must be null or integers.');
+        }
+
+        if (
+            $type === self::RANGE &&
+            (
+                ($startOffset !== null && !is_int($startOffset) && !is_string($startOffset)) ||
+                ($endOffset !== null && !is_int($endOffset) && !is_string($endOffset))
+            )
+        ) {
+            throw new InvalidArgumentException('Frame offsets for RANGE must be null, integers or interval strings.');
+        }
+
+        $this->frame = [
+            'type' => $type,
+            'start' => [
+                'offset' => $startOffset,
+                'direction' => $startDirection,
+            ],
+        ];
+
+        if (func_num_args() > 3) {
```

Comment:
What if the user actaully passed all the arguments, but with the null defaults? Would this break?

Reply:
`func_num_args()` returns the actual number of arguments passed. Even though there are defaults set, this won't include them. This way we can tell if the user wants 'BETWEEN ... FOLLOWING' generated or if they want the default implementation without writing that out.

Reply:
My question is what if the user actually pass more then 3 arguments like this `->frame($a, $b, $c, null, 0)`
Won't this cause a problem aftewards as the offset will be `null`?

Reply:
null is a valid offset argument.